static/app/views/dashboards/widgetCard/autoSizedText.tsx

```diff
@@ -0,0 +1,177 @@
+      setCalculationCount(0);
+      setFontSizeLowerBound(minFontSize);
+      setFontSizeUpperBound(maxFontSize);
+
+      // Update state, trigger a re-render
+      setParentHeight(parentDimensions.height);
+      setParentWidth(parentDimensions.width);
+    }
+  }, RESIZE_DEBOUNCE);
+
+  useResizeObserver({
+    ref: parentRef,
+    onResize,
+  });
+
+  useLayoutEffect(() => {
+    if (calculationCount > calculationCountLimit) {
+      // Exceeded the iteration count. This should be unlikely! If it happens, abandon
+      return;
+    }
+
+    const parentDimensions = getElementDimensions(parentRef.current);
+    const childDimensions = getElementDimensions(childRef.current);
+
+    if (!parentDimensions || !childDimensions) {
+      // Refs are not ready or cannot be measured, abandon
+      return;
+    }
+
+    // Calculate the width and height disparity between the child and parent. A disparity of 0 means they're the same size.
+    const widthDisparity = calculateDimensionDisparity(
+      childDimensions,
+      parentDimensions,
+      'width'
+    );
+
+    const heightDisparity = calculateDimensionDisparity(
+      childDimensions,
+      parentDimensions,
+      'height'
+    );
+
+    const childFitsIntoParent =
+      childDimensions.width <= parentDimensions.width &&
+      childDimensions.height <= parentDimensions.height;
+
+    if (
+      childFitsIntoParent &&
+      (widthDisparity <= MAXIMUM_DISPARITY || heightDisparity <= MAXIMUM_DISPARITY)
+    ) {
+      // The child fits completely into the parent _and_ at least one dimension is very similar to the parent size (i.e., it fits nicely in the parent). Abandon, we're done!
+      return;
+    }
+
+    if (
+      childDimensions.width > parentDimensions.width ||
+      childDimensions.height > parentDimensions.height
+    ) {
+      // The element is bigger than the parent, scale down
+      const newFontSize = (fontSizeLowerBound + fontSize) / 2;
+
+      startTransition(() => {
+        setCalculationCount(previousCalculationCount => previousCalculationCount + 1);
+        setFontSizeUpperBound(fontSize);
+        setFontSize(newFontSize);
+      });
+    } else if (
+      childDimensions.width < parentDimensions.width ||
+      childDimensions.height < parentDimensions.height
+    ) {
+      // The element is too small, scale up
+      const midpoint = (fontSizeUpperBound + fontSize) / 2;
+
+      startTransition(() => {
+        setCalculationCount(previousCalculationCount => previousCalculationCount + 1);
+        setFontSizeLowerBound(fontSize);
+        setFontSize(midpoint);
+      });
+    }
+  }, [
+    fontSize,
+    minFontSize,
+    fontSizeLowerBound,
+    maxFontSize,
+    fontSizeUpperBound,
+    calculationCount,
+    calculationCountLimit,
+    parentHeight,
+    parentWidth,
+  ]);
+
+  return (
+    <ParentSizeMimic ref={parentRef}>
+      <SizedChild ref={childRef} style={{fontSize}}>
```

Comment:
There is constraint here, which I think could be avoided. Since the autosize component does not implement the HTMLElement interface, it means the callers cannot override anything about this intermediary wrapper - we have something that can no longer be controller (I dont know if this will create problems in practice, but I'd rather have more flexibility in case we need it and defer this to whoever is using the component). 

There are two ways to solve it, one is to either implement the interface and spread the props to our ParentSizeMimic component, this way we get full access and can override the wrapper, but the second and I think more interesting option is to **not** render the wrapper at all and instead rely on childRef.parentElement reference to do all the calculations (this way you can even remove the parent ref logic and things like if(!parentRef ||! childRef) logic